### PR TITLE
Fix for asa_config #25319

### LINKED ISSUE #25319: asa_config issue
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

asa_config
##### ANSIBLE VERSION

```
[root@1be66e998fd3 module_utils]# ansible --version
ansible 2.4.0
  config file =
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
```

##### CONFIGURATION

n/a
##### OS / ENVIRONMENT

ASA 9.4(3)
also enable following on ASA, so user can be in enable mode after login.

##### SUMMARY

run asa_config with error

##### STEPS TO REPRODUCE

ansible-playbook -i labhost fwobject.yaml -vvvv

```yaml
- name: Cisco ASA access-list config
  connection: local
  hosts: testfw
  gather_facts: false
  vars:
    cli:
      username: "xxx"
      password: "xxx"
      host: "10.0.203.251"
      authorize: no
      transport: cli

  tasks:
    - name: create object group
      asa_config:
        lines:
          - network-object host 10.1.0.1
          - network-object host 10.1.0.2
          - network-object host 10.1.0.3
        parents: ['object-group network dummy-group']
        provider: "{{ cli }}"
```

##### ACTUAL RESULTS

[root@1be66e998fd3 work]# ansible-playbook -i labhost fwobject.yaml -vvvv
ansible-playbook 2.4.0
  config file =
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
No config file found; using defaults
setting up inventory plugins
Parsed /work/labhost inventory source with ini plugin
Loading callback plugin default of type stdout, v2.0 from /usr/lib/python2.7/site-packages/ansible/plugins/callback/__init__.pyc

PLAYBOOK: fwobject.yaml *********************************************************************************************************************
1 plays in fwobject.yaml

PLAY [Cisco ASA access-list config] *********************************************************************************************************
META: ran handlers

TASK [create object group] ******************************************************************************************************************
task path: /work/fwobject.yaml:14
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/basic.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/asa.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/network.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/netcfg.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/_text.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/pycompat24.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/six/__init__.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/six/_six.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/shell.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/netcli.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/network_common.py
Using module file /usr/lib/python2.7/site-packages/ansible/modules/network/asa/asa_config.py
<10.0.203.251> ESTABLISH LOCAL CONNECTION FOR USER: root
<10.0.203.251> EXEC /bin/sh -c 'echo ~ && sleep 0'
<10.0.203.251> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1496437577.05-213904690690345 `" && echo ansible-tmp-1496437577.05-213904690690345="` echo /root/.ansible/tmp/ansible-tmp-1496437577.05-213904690690345 `" ) && sleep 0'
<10.0.203.251> PUT /tmp/tmpzHofGc TO /root/.ansible/tmp/ansible-tmp-1496437577.05-213904690690345/asa_config.py
<10.0.203.251> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1496437577.05-213904690690345/ /root/.ansible/tmp/ansible-tmp-1496437577.05-213904690690345/asa_config.py && sleep 0'
<10.0.203.251> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1496437577.05-213904690690345/asa_config.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1496437577.05-213904690690345/" > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_xPseWY/ansible_module_asa_config.py", line 349, in <module>
    main()
  File "/tmp/ansible_xPseWY/ansible_module_asa_config.py", line 341, in main
    run(module, result)
  File "/tmp/ansible_xPseWY/ansible_module_asa_config.py", line 271, in run
    config = get_config(module)
  File "/tmp/ansible_xPseWY/ansible_module_asa_config.py", line 251, in get_config
    contents = module.config.get_config(include=include)
  File "/tmp/ansible_xPseWY/ansible_modlib.zip/ansible/module_utils/network.py", line 98, in get_config
  File "/tmp/ansible_xPseWY/ansible_modlib.zip/ansible/module_utils/asa.py", line 107, in get_config
  File "/tmp/ansible_xPseWY/ansible_modlib.zip/ansible/module_utils/shell.py", line 284, in <lambda>
  File "/tmp/ansible_xPseWY/ansible_modlib.zip/ansible/module_utils/shell.py", line 274, in execute
  File "/tmp/ansible_xPseWY/ansible_modlib.zip/ansible/module_utils/shell.py", line 263, in to_command
TypeError: __init__() takes exactly 3 arguments (2 given)

fatal: [10.0.203.251]: FAILED! => {
    "changed": false,
    "failed": true,
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_xPseWY/ansible_module_asa_config.py\", line 349, in <module>\n    main()\n  File \"/tmp/ansible_xPseWY/ansible_module_asa_config.py\", line 341, in main\n    run(module, result)\n  File \"/tmp/ansible_xPseWY/ansible_module_asa_config.py\", line 271, in run\n    config = get_config(module)\n  File \"/tmp/ansible_xPseWY/ansible_module_asa_config.py\", line 251, in get_config\n    contents = module.config.get_config(include=include)\n  File \"/tmp/ansible_xPseWY/ansible_modlib.zip/ansible/module_utils/network.py\", line 98, in get_config\n  File \"/tmp/ansible_xPseWY/ansible_modlib.zip/ansible/module_utils/asa.py\", line 107, in get_config\n  File \"/tmp/ansible_xPseWY/ansible_modlib.zip/ansible/module_utils/shell.py\", line 284, in <lambda>\n  File \"/tmp/ansible_xPseWY/ansible_modlib.zip/ansible/module_utils/shell.py\", line 274, in execute\n  File \"/tmp/ansible_xPseWY/ansible_modlib.zip/ansible/module_utils/shell.py\", line 263, in to_command\nTypeError: __init__() takes exactly 3 arguments (2 given)\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 0
}
	to retry, use: --limit @/work/fwobject.retry

PLAY RECAP **********************************************************************************************************************************
10.0.203.251               : ok=0    changed=0    unreachable=0    failed=1

```

```

### PR DESCRIPTION
**Pull request by ogenstad**
##### SUMMARY
This would fix #25319, however more things will need to be done to restore the full functionality of the ASA modules.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - asa_config

##### ANSIBLE VERSION
```
ansible 2.4.0 (asa_config 2c86bd2b72) last updated 2017/06/14 09:43:22 (GMT +200)
  config file =
  configured module search path = [u'/Users/patrick/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/patrick/src/forks/ansible/lib/ansible
  executable location = /Users/patrick/src/forks/ansible/bin/ansible
  python version = 2.7.12 (default, Dec  5 2016, 11:55:02) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

##### ADDITIONAL INFORMATION
I don't like the way I solved this, and this is mostly pushed to show what could be done. The changes to shell.py have been done so that it shouldn't impact any other module. However there's a good chance that other modules using that code doesn't work now either.

### CODE REVIEW

**Comment by gundalow** [comment]
I've added this to the agenda for the Network IRC meeting

File: lib/ansible/module_utils/shell.py [hunk 1]
```diff
@@ -256,22 +256,31 @@ def to_command(self, obj):
...
-            ))
+            if module:
+                transform = ComplexDict(dict(
+                    command=dict(key=True),
+                    prompt=dict(),
+                    answer=dict(),
+                    sendonly=dict(default=False)
+                ), module)
+            else:
+                transform = ComplexDict(dict(
```
Comment (Qalthos) [inline lib/ansible/module_utils/shell.py:267]: I don't think there's any reason to keep this path, ComplexDict needs a module parameter

**Review by Qalthos — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/shell.py (hunk 1 above)
Comment (ogenstad) [reply lib/ansible/module_utils/shell.py:267]: No, this was just to get it working and not touch anything for other modules. But I agree that this and other things needs to be changed.

**Comment by gundalow** [comment]
Discussed in IRC Meeting https://meetbot.fedoraproject.org/ansible-meeting/2017-06-14/network_working_group.2017-06-14-16.02.log.html

**Pull request closed without merging**